compiler/ir/backend.jvm/src/org/jetbrains/kotlin/backend/jvm/codegen/IrSourceCompilerForInline.kt

```diff
@@ -127,8 +127,12 @@ class IrSourceCompilerForInline(
         return SMAPAndMethodNode(node!!, SMAP(classCodegen.getOrCreateSourceMapper().resultMappings))
     }
 
-    override fun generateLambdaBody(lambdaInfo: ExpressionLambda): SMAPAndMethodNode =
-        makeInlineNode((lambdaInfo as IrExpressionLambdaImpl).function, codegen.classCodegen, true)
+    override fun generateLambdaBody(lambdaInfo: ExpressionLambda): SMAPAndMethodNode {
+        val function = (lambdaInfo as IrExpressionLambdaImpl).function
+        return makeInlineNode(
+            function, if (lambdaInfo.isCrossInline) codegen.classCodegen else FakeClassCodegen(function, codegen.classCodegen), true
```

Comment:
This does not seem right.
1. Why the exception for crossinline lambdas?
2. This affects more than the source maps, e.g. anonymous objects defined in (non-crossinline) inline lambdas would no longer be listed in InnerClasses of the class where the lambda is declared.
Also note that #3056 removes FakeClassCodegen.

Reply:
My strategy was simple: in a debug mode, stepping through source map create points side-by-side, i.e., comparing old and new backends. Then, spotted here: https://github.com/JetBrains/kotlin/blob/master/compiler/backend/src/org/jetbrains/kotlin/codegen/inline/SourceCompilerForInline.kt#L153, an override of `generateLambdaBody` in old backend, which in turn goes to `generateMethodBody` where `FakeMemberCodegen` (a counterpart of `FakeClassCodegen`) is always used.

> 1. Why the exception for crossinline lambdas?

From here: https://github.com/JetBrains/kotlin/blob/master/compiler/backend/src/org/jetbrains/kotlin/codegen/inline/SourceCompilerForInline.kt#L152 the control goes down to here: https://github.com/JetBrains/kotlin/blob/master/compiler/backend/src/org/jetbrains/kotlin/codegen/context/InlineLambdaContext.kt#L29 where, for crossinline, the original context was used. Well, it has other conditions too, and I think those other conditions are handled in `IrSourceCompilerForInline#makeInlineNode` I think.

> 2. ... anonymous objects defined in (non-crossinline) inline lambdas would no longer be listed in InnerClasses of the class where the lambda is declared.

To be honest, I can't follow this fully. I wonder, though, we don't have good test coverage for that scenario (since no test is broken)?

> 3. FakeClassCodegen here loses the link to the parent function ... (this is probably why you needed the other change to avoid exceptions).

You're right. Delegating parent function (in FakeClassCodegen) resolves it. Also, not needed to override a map function in identical source mapper.

> Also note that #3056 removes FakeClassCodegen.

Thanks for the heads-up. I've tried to rewrite this change without using FakeClassCodegen, like that PR does, but not successful yet. :\

Reply:
>To be honest, I can't follow this fully. I wonder, though, we don't have good test coverage for that scenario (since no test is broken)?

We *kind of* do. The IR backend generates objects in lambdas differently from the old backend: in the old backend, objects capture the lambda itself, then use that "outer this" to access all captured variables; in the new backend, objects capture everything they need directly. Because of this, the old backend needs to regenerate those objects after inlining the lambdas (since there's no longer a lambda object) while the new backend does not. This leads to visible differences in class names, which makes tests for InnerClass attributes fail on JVM_IR.

(This is also why it's fine to use a completely fake codegen in the old backend, where the classes are going to be removed anyway and replaced with regenerated copies, but not in the old backend, where you need to generate the original object correctly.)

Reply:
Alright, I have to admit a special handling of `crossinline` at this point is quite awkward, but was not able to understand a new test failure. See my other comment inlined to that test. Perhaps, it's still related to InnerClasses...?

Also, my new change is... yeah it's still another variant of FakeClassCodegen, but in preparation of sunset of it at #3056 . Still can't grasp the ideas of that PR and rewrite things on top of it...

Reply:
1. special handling of `crossinline` is gone.
3. link(s) to parent function (as well as parent class) are handled.

Remaining one is 2. Not sure parent class is the right term here; using the current context's IR class is same as using the enclosing class or maybe not? Linking it to the new ClassCodegen is not enough to keep InnerClasses attributes for that enclosing class?

Anyhow, the latest commit is a bit different from the initial one, so PTAL.